File: leetcode/0432-all-oone-data-structure/Solution.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class AllOne {
    private class Node {
        int count;
        Set<String> keys = new HashSet<>();
        Node prev, next;
        Node(int count) { this.count = count; }
    }

    private final Node head = new Node(0);
    private final Node tail = new Node(0);
    private final Map<String, Integer> keyCount = new HashMap<>();
    private final Map<String, Node> keyNode = new HashMap<>();

    public AllOne() {
        head.next = tail;
        tail.prev = head;
    }

    private Node insertAfter(Node node, int count) {
        Node newNode = new Node(count);
        newNode.prev = node;
        newNode.next = node.next;
        node.next.prev = newNode;
        node.next = newNode;
        return newNode;
    }

    private Node insertBefore(Node node, int count) {
        Node newNode = new Node(count);
        newNode.next = node;
        newNode.prev = node.prev;
        node.prev.next = newNode;
        node.prev = newNode;
        return newNode;
    }

    private void remove(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    public void inc(String key) {
        if (!keyCount.containsKey(key)) {
            keyCount.put(key, 1);
            Node node;
            if (head.next == tail || head.next.count != 1) {
                node = insertAfter(head, 1);
            } else {
                node = head.next;
            }
            node.keys.add(key);
            keyNode.put(key, node);
            return;
        }

        Node curNode = keyNode.get(key);
        int newCount = keyCount.get(key) + 1;
        keyCount.put(key, newCount);
        curNode.keys.remove(key);

        Node newNode;
        if (curNode.next == tail || curNode.next.count != newCount) {
            newNode = insertAfter(curNode, newCount);
        } else {
            newNode = curNode.next;
        }
        newNode.keys.add(key);
        keyNode.put(key, newNode);

        if (curNode.keys.isEmpty()) remove(curNode);
    }

    public void dec(String key) {
        if (!keyCount.containsKey(key)) return;

        Node curNode = keyNode.get(key);
        int curCount = keyCount.get(key);
        curNode.keys.remove(key);

        if (curCount == 1) {
            keyCount.remove(key);
            keyNode.remove(key);
        } else {
            int newCount = curCount - 1;
            keyCount.put(key, newCount);
            Node newNode;
            if (curNode.prev == head || curNode.prev.count != newCount) {
                newNode = insertBefore(curNode, newCount);
            } else {
                newNode = curNode.prev;
            }
            newNode.keys.add(key);
            keyNode.put(key, newNode);
        }

        if (curNode.keys.isEmpty()) remove(curNode);
    }

    public String getMaxKey() {
        if (tail.prev == head) return "";
        return tail.prev.keys.iterator().next();
    }

    public String getMinKey() {
        if (head.next == tail) return "";
        return head.next.keys.iterator().next();
    }
}

/**
 * Your AllOne object will be instantiated and called as such:
 * AllOne obj = new AllOne();
 * obj.inc(key);
 * obj.dec(key);
 * String param_3 = obj.getMaxKey();
 * String param_4 = obj.getMinKey();
 */

File: leetcode/0432-all-oone-data-structure/solution.py
class _Node:
    __slots__ = ("count", "keys", "prev", "next")

    def __init__(self, count=0):
        self.count = count
        self.keys = set()
        self.prev = None
        self.next = None


class AllOne:
    def __init__(self):
        # sentinels: head.next is the lowest-count bucket, tail.prev is the highest-count bucket
        self.head = _Node()
        self.tail = _Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.key_count = {}  # key -> current count
        self.key_node = {}   # key -> bucket node currently holding it

    def _insert_after(self, node, count):
        new_node = _Node(count)
        new_node.prev = node
        new_node.next = node.next
        node.next.prev = new_node
        node.next = new_node
        return new_node

    def _insert_before(self, node, count):
        new_node = _Node(count)
        new_node.next = node
        new_node.prev = node.prev
        node.prev.next = new_node
        node.prev = new_node
        return new_node

    def _remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def inc(self, key: str) -> None:
        if key not in self.key_count:
            self.key_count[key] = 1
            if self.head.next is self.tail or self.head.next.count != 1:
                node = self._insert_after(self.head, 1)
            else:
                node = self.head.next
            node.keys.add(key)
            self.key_node[key] = node
            return

        cur_node = self.key_node[key]
        new_count = self.key_count[key] + 1
        self.key_count[key] = new_count
        cur_node.keys.remove(key)

        if cur_node.next is self.tail or cur_node.next.count != new_count:
            new_node = self._insert_after(cur_node, new_count)
        else:
            new_node = cur_node.next
        new_node.keys.add(key)
        self.key_node[key] = new_node

        if not cur_node.keys:
            self._remove(cur_node)

    def dec(self, key: str) -> None:
        if key not in self.key_count:
            return

        cur_node = self.key_node[key]
        cur_count = self.key_count[key]
        cur_node.keys.remove(key)

        if cur_count == 1:
            del self.key_count[key]
            del self.key_node[key]
        else:
            new_count = cur_count - 1
            self.key_count[key] = new_count
            if cur_node.prev is self.head or cur_node.prev.count != new_count:
                new_node = self._insert_before(cur_node, new_count)
            else:
                new_node = cur_node.prev
            new_node.keys.add(key)
            self.key_node[key] = new_node

        if not cur_node.keys:
            self._remove(cur_node)

    def getMaxKey(self) -> str:
        if self.tail.prev is self.head:
            return ""
        return next(iter(self.tail.prev.keys))

    def getMinKey(self) -> str:
        if self.head.next is self.tail:
            return ""
        return next(iter(self.head.next.keys))


# Your AllOne object will be instantiated and called as such:
# obj = AllOne()
# obj.inc(key)
# obj.dec(key)
# param_3 = obj.getMaxKey()
# param_4 = obj.getMinKey()
